src/sietch/distributed_dune.ml

```diff
@@ -95,45 +135,34 @@ let index_put t name hash lines =
     (name, Digest.to_string hash, lines)
   |> async_ok
 
-let distribute ({ cache; _ } as t) key (metadata : Cache.Local.Metadata_file.t)
-    =
-  let f () =
-    let* () =
-      match metadata.contents with
-      | Files files ->
-        let insert_file { Cache.File.digest; _ } =
-          let path = Local.file_path cache digest in
-          let insert stats input =
-            let executable = stats.Unix.st_perm land 0o100 > 0 in
-            let* contents = Async.Reader.contents input |> async_ok in
-            block_put t digest executable contents
-          in
-          let stats = Path.stat path in
-          let* upload =
-            if stats.st_size < 4096 then
-              Async.Deferred.Result.return true
-            else
-              block_has t digest
-          in
-          if upload then
-            let path = Path.to_string path in
-            let () = debug [ Pp.textf "distribute %S" path ] in
-            Async.Reader.with_file path ~f:(insert stats)
-          else
-            Async.Deferred.Result.return ()
-        in
-        let ( let* ) = Async.Deferred.( >>= ) in
-        let* results =
-          Async.Deferred.List.map ~how:(`Max_concurrent_jobs 8) ~f:insert_file
-            files
-        in
-        let ( let* ) = Async.Deferred.Result.( >>= ) in
-        let* (_ : unit list) = results |> Result.List.all |> Async.return in
-        Async.Deferred.Result.return ()
-      | Value _ -> Async.Deferred.Result.fail "ignoring Jenga value"
+let metadata_put t hash (metadata : Cache.Local.Metadata_file.t) =
+  let* targets =
+    let ( let* ) = Async.( >>= ) in
+    let* init =
+      let+ client, uri = client_and_uri t hash in
+      Clients.set Clients.empty uri client
+    in
+    let f init { Cache.File.digest; _ } =
+      match init with
+      | Result.Ok init ->
+        let+ client, uri = client_and_uri t digest in
+        Clients.set init uri client
+      | Result.Error e -> Async.Deferred.Result.fail e
     in
-    block_put t key false (Cache.Local.Metadata_file.to_string metadata)
+    match metadata.contents with
+    | Value v ->
+      let+ client, uri = client_and_uri t v in
+      Clients.set Clients.empty uri client
+    | Files files -> Async.Deferred.List.fold ~f ~init files
   in
+  let f (_, client) =
+    Async.Rpc.Rpc.dispatch_exn Dune_distributed_storage.Rpc.metadata_put client
+      (Digest.to_string hash, Cache.Local.Metadata_file.to_string metadata)
+  in
+  Async.Deferred.List.iter ~f (Clients.to_list targets) |> async_ok
```

Comment:
This function is pretty hard to review without a short spec for its intended behaviour.